### PR TITLE
Introduce local scheduler heartbeats which carry load information.

### PR DESCRIPTION
**Pull request by robertnishihara**
This implements `local_scheduler_table_subscribe`, which is called in `global_scheduler.c` to receive heartbeats from the local schedulers.

This implements `local_scheduler_table_send_info`, which is called on a timer in `photon_scheduler.c` to send heartbeats.

The information being sent is the `local_scheduler_info` struct defined in `local_scheduler_table.h`.

### CODE REVIEW

**Review by pcmoritz — APPROVED** [review]

File: src/common/state/redis.c [hunk 1]
```diff
@@ -1081,6 +1082,82 @@ void redis_db_client_table_subscribe(table_callback_data *callback_data) {
...
+  REDIS_CALLBACK_HEADER(db, callback_data, r);
+
+  redisReply *reply = r;
+  CHECK(reply->type == REDIS_REPLY_ARRAY);
+  CHECK(reply->elements == 3);
+  redisReply *message_type = reply->element[0];
+  LOG_DEBUG("Local scheduer table subscribe callback, message %s",
+            message_type->str);
+
+  if (strcmp(message_type->str, "message") == 0) {
```
Comment (atumanov) [inline src/common/state/redis.c]: when the comparison is with a fixed-length string of known size, it is always preferred to use strncmp(str1, str2, size)

File: src/common/state/redis.c [hunk 2]
```diff
@@ -1081,6 +1082,82 @@ void redis_db_client_table_subscribe(table_callback_data *callback_data) {
...
+}
+
+void redis_local_scheduler_table_send_info_callback(redisAsyncContext *c,
+                                                    void *r,
+                                                    void *privdata) {
+  REDIS_CALLBACK_HEADER(db, callback_data, r);
+
+  redisReply *reply = r;
+  CHECK(reply->type == REDIS_REPLY_INTEGER);
+  LOG_DEBUG("%lld subscribers received this publish.\n", reply->integer);
```
Comment (atumanov) [inline src/common/state/redis.c]: we should probably go with PRId64 here

File: src/common/state/local_scheduler_table.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#ifndef LOCAL_SCHEDULER_TABLE_H
+#define LOCAL_SCHEDULER_TABLE_H
+
+#include "db.h"
+#include "table.h"
+
+typedef struct {
+  int task_queue_length;
+  int available_workers;
```
Comment (atumanov) [inline src/common/state/local_scheduler_table.h]: I would throw in total_workers as well, but we can make extensions a separate PR.
This is the nice thing about putting everything in a struct 👍 . Easy to extend.

File: src/photon/photon_scheduler.h [hunk 4]
```diff
@@ -4,6 +4,9 @@
 #include "task.h"
 #include "event_loop.h"
 
+/* The duration between local scheduler heartbeats. */
+#define LOCAL_SCHEDULER_HEARTBEAT_TIMEOUT_MILLISECONDS 1000
```
Comment (atumanov) [inline src/photon/photon_scheduler.h]: one heartbeat per second is going to be ok for checking liveness, but too infrequent for available_workers information to be actionable. Part of the issue here is that we should really be running these updates in a separate thread with read-only access to the queue information.

**Review by atumanov — APPROVED** [review]
nice.
